### PR TITLE
p2sh psbt convenience methods

### PR DESCRIPTION
**Pull request by mflaxman**
UPDATE: Finally ready to merge!

Things missing (see `TODO`s in code):
- [x] Change detection (https://github.com/buidl-bitcoin/buidl-python/pull/81/commits/7c91e98b83c5cfc671d7d3d46076819e4d59a3d7 and earlier)
- [x] Add support for non-lexicographical ordering of child pubkeys
- [x] Better variable names. UPDATE: done in https://github.com/buidl-bitcoin/buidl-python/pull/81/commits/531bb1ab8e1791cbbbd482bc4fc613acc55ae7cb
- [x] Add optional `target_addr` argument to `create_p2sh_multisig` method for safety (done in https://github.com/buidl-bitcoin/buidl-python/pull/81/commits/3b939190d1113c7d494609608407ea7cf85548f8)

No
* ~~Have option to only require the prev_hash/idx/amount and not the full tx hex for all UTXOs?~~ This isn't compatible with coldcard, so *not* going to do this for now.

### CODE REVIEW

**Comment by mflaxman** [comment]
Would `input_dicts` be better than `inputs_dict_list`? Similarly for `output_dicts` vs `outputs_dict_list`?

File: buidl/psbt_helper.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
+from collections import defaultdict
```
Comment (dhruvbansal) [inline buidl/psbt_helper.py:1]: After some discussion with Chris, I think we have a workaround and we can probably remove this entire file/method from buidl. 

I think that will be *much* cleaner from the build side!

File: buidl/script.py [hunk 2]
```diff
@@ -369,34 +371,92 @@ def convert(cls, raw_redeem_script):
...
+            calculated_addr = to_return.address(network=expected_addr_network)
+            if expected_addr != calculated_addr:
+                raise ValueError(
+                    f"Expected address {expected_addr} but calculated {calculated_addr}"
+                )
+
+        return to_return
+
+    @classmethod
+    def create_p2sh_multisig_unsorted(
```
Comment (dhruvbansal) [inline buidl/script.py:422]: Because we can lose the psbt helper, I think we can lose this method too.  Thank god.

File: buidl/script.py [hunk 3]
```diff
@@ -369,34 +371,92 @@ def convert(cls, raw_redeem_script):
...
-            raise ValueError(f"Invalid m-of-n: {quorum_m}-of-{len(pubkey_hex_list)}")
+        if quorum_m < 1 or quorum_m > len(pubkey_hexes):
+            raise ValueError(f"Invalid m-of-n: {quorum_m}-of-{len(pubkey_hexes)}")
 
         commands = [number_to_op_code(quorum_m)]
 
         if sort_keys:
-            pubkey_hexes = sorted(pubkey_hex_list)
+            pubkey_hexes = sorted(pubkey_hexes)
         else:
```
Comment (dhruvbansal) [inline buidl/script.py:399]: Is this `else` clause necessary?

File: buidl/script.py [hunk 4]
```diff
@@ -369,34 +371,92 @@ def convert(cls, raw_redeem_script):
...
+        )
+
+    def get_quorum(self):
+        """
+        Return the m-of-n of this multisig, as in 2-of-3 or 3-of-5
+        """
+        if self.commands[-1] != 174:
+            raise ValueError(f"Not OP_CHECKMULTISIG: {self}")
+        quorum_m = op_code_to_number(self.commands[0])
+        quorum_n = len(self.commands) - 3
```
Comment (dhruvbansal) [inline buidl/script.py:452]: A comment about where the `3` comes from would be nice (I think it's `OP_CHECKMULTISIG`, `quorum_m` and the off-by-one issue?)

File: buidl/script.py [hunk 5]
```diff
@@ -369,34 +371,92 @@ def convert(cls, raw_redeem_script):
...
+        return quorum_m, quorum_n
+
+    def signing_pubkeys(self):
+        """
+        The pubkeys needed to sign this transaction, typically children derived from xpubs
+        """
+        return self.commands[1:-2]
```
Comment (dhruvbansal) [inline buidl/script.py:459]: Do you need a similar guard as you had in `get_quorum` to ensure this is a multisig?

**Review by dhruvbansal — COMMENTED** [review]
The P2SH script implementation looks really good!  I think we should remove the `create_ps2sh_multisig_psbt` from buidl.

File: buidl/psbt.py [hunk 6]
```diff
@@ -190,12 +196,23 @@ def parse(cls, key, s, network=None):
         return hd_key
 
     @classmethod
-    def from_hd_priv(cls, hd_priv, path):
-        hd_key = hd_priv.traverse(path).pub
+    def from_hd_pub(cls, child_hd_pub, fingerprint_hex, root_path):
```
[the remark this answers is not in the document]
Comment (mflaxman) [reply buidl/psbt.py]: Good call, fixed in https://github.com/buidl-bitcoin/buidl-python/pull/81/commits/ab0427b9966148f0381574c1e21febeadb2b37ee

File: buidl/psbt_helper.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+                base_path=base_path,
+            )
+            if child_path:
+                if base_path.count("/") != xpub_obj.depth:
+                    msg = f"base_path {base_path} depth != {xpub_obj.depth} for {xpub_obj}"
+                    raise ValueError(msg)
+                return xpub_obj.traverse(child_path)
+
+
+def create_ps2sh_multisig_psbt(
```
[the remark this answers is not in the document]
Comment (mflaxman) [reply buidl/psbt_helper.py]: Oof 🤦.

Fixed in https://github.com/buidl-bitcoin/buidl-python/pull/81/commits/6e6b5fc9f83c89225324f9430691f2c1cedb5939

File: buidl/script.py (hunk 3 above)
Comment (mflaxman) [reply buidl/script.py]: Good catch, not sure how that got weird. Fixed in https://github.com/buidl-bitcoin/buidl-python/pull/81/commits/a4717e714250a23139e3e1c150ded31ff30f276e

File: buidl/script.py (hunk 4 above)
Comment (mflaxman) [reply buidl/script.py:451]: Correct! Added a comment to make it unambiguous in https://github.com/buidl-bitcoin/buidl-python/pull/81/commits/5c3bd5ecff35bf0b38e2523c60d6b218e11c61ea

File: buidl/script.py (hunk 5 above)
Comment (mflaxman) [reply buidl/script.py:463]: The `RedeemScript` subclass represents a RedeemScript for `p2sh`, but I added a safety check to be extra sure it isn't misused. I like having lots of defensive checks that should never hit; if anything goes wrong you at least know right away where things are off.

Added in https://github.com/buidl-bitcoin/buidl-python/pull/81/commits/c281b3e7c105097c7d74d99758bc079652285eee

**Comment by mflaxman** [comment]
@dhruvbansal thanks for the code review!

It makes sense if you want to to do your own custom thing, and `buidl` allows you to do it however you like.

I got rid of the complicated unsorted `RedeemScript` helper method, as that's pretty niche. But, if you want to create a `RedeemScript` from out of order pubkeys the existing method allows it natively (and has a safety check where you can specify the `expected_addr`).

I left in the general `create_p2sh_multisig_psbt()` helper method because I think the functionality is important for building on top of (and as a bonus I like adding test coverage to `buidl` wherever possible). That said, it's in a separate `psbt_helper.py1 file so it should keep the `psbt.py` file as easy to read as possible. I'd like to one day add similar methods for `p2wsh`/`p2sh-p2wsh`, so that users can even run a multisig coordinator (and signer) without even needing a bitcoind node.

Does that work/make sense for you? Changes are easiest to see in https://github.com/buidl-bitcoin/buidl-python/pull/81/commits/5cae54f7c870926876fc7b274e9737ae65396d38 .

**Comment by mflaxman** [comment]
Now live on pypi:
https://pypi.org/project/buidl/0.2.21/

**Pull request merged**